clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java

```diff
@@ -0,0 +1,274 @@
+            mockLogContext,
+            mockTime,
+            heartbeatInterval,
+            retryBackoffMs,
+            retryBackoffMaxMs);
+        heartbeatRequestManager = createManager();
+
+        when(mockMembershipManager.state()).thenReturn(state);
+        mockTime.sleep(100); // time elapsed < heartbeatInterval, no heartbeat should be sent
+        NetworkClientDelegate.PollResult result = heartbeatRequestManager.poll(mockTime.milliseconds());
+        assertEquals(0, result.unsentRequests.size());
+
+        if (mockMembershipManager.shouldSendHeartbeat()) {
+            assertEquals(heartbeatInterval - 100, result.timeUntilNextPollMs);
+        } else {
+            assertEquals(Long.MAX_VALUE, result.timeUntilNextPollMs);
+        }
+    }
+
+    @Test
+    public void testNoCoordinator() {
+        when(mockCoordinatorRequestManager.coordinator()).thenReturn(Optional.empty());
+        NetworkClientDelegate.PollResult result = heartbeatRequestManager.poll(mockTime.milliseconds());
+
+        assertEquals(Long.MAX_VALUE, result.timeUntilNextPollMs);
+        assertEquals(0, result.unsentRequests.size());
+    }
+
+    @ParameterizedTest
+    @MethodSource("errorProvider")
+    public void testHeartbeatResponseOnErrorHandling(final Errors error, final boolean isFatal) {
+        heartbeatRequestState = new HeartbeatRequestManager.HeartbeatRequestState(
+            mockLogContext,
+            mockTime,
+            heartbeatInterval,
+            retryBackoffMs,
+            retryBackoffMaxMs,
+            0);
+        errorEventHandler = mock(ErrorEventHandler.class);
+        when(mockMembershipManager.state()).thenReturn(MemberState.RECONCILING);
+        heartbeatRequestManager = createManager();
+
+        mockTime.sleep(1001);
+        NetworkClientDelegate.PollResult result = heartbeatRequestManager.poll(mockTime.milliseconds());
+        assertEquals(1, result.unsentRequests.size());
+
+        // manually completing the response to test error handling
+        ClientResponse response = createHeartbeatResponse(
+            result.unsentRequests.get(0),
+            error);
+        result.unsentRequests.get(0).future().complete(response);
+        ConsumerGroupHeartbeatResponse mockResponse = (ConsumerGroupHeartbeatResponse) response.responseBody();
+
+        short errorCode = error.code();
+        if (errorCode == Errors.NONE.code()) {
+            verify(errorEventHandler, never()).handle(any());
+            verify(mockMembershipManager).updateState(mockResponse.data());
+            assertEquals(heartbeatInterval, heartbeatRequestState.nextHeartbeatMs(mockTime.milliseconds()));
+        } else if (isFatal) {
+            verify(errorEventHandler).handle(any());
+            assertEquals(retryBackoffMs, heartbeatRequestState.nextHeartbeatMs(mockTime.milliseconds()));
+            if (errorCode == Errors.UNRELEASED_INSTANCE_ID.code()) {
```

Comment:
If we agree on the [comment](https://github.com/apache/kafka/pull/14364/files#r1335304701) regarding  missing transitions when handling fatal errors, I expect this will be updated to reflect align with it and check transition to FAILED on all fatal errors other than the 2 fencing ones.